File: crates/nxvim-core/src/editor/explorer.rs
//! The in-window directory listing — nxvim's file explorer (vim's netrw).
//!
//! Opening a directory (`nxvim .`, `:e somedir`, `<CR>` on a listed sub-directory)
//! builds a read-only [`Buffer`] whose lines are the directory's entries (see
//! [`Buffer::from_dir`]). Such a buffer carries `dir: Some(path)`, so
//! [`Buffer::read_only`] is true and edits are refused at the
//! [`modifiable`](Editor::modifiable) chokepoints — the listing stays a faithful
//! picture of the filesystem. It is otherwise an **ordinary buffer in a window**
//! (vim's netrw model): `j`/`k`/`gg`/`G`/`/`/`:` are ordinary normal-mode motions,
//! and its two activation keys — `<CR>` (open) and `-` (parent), via
//! [`apply_explorer_action`](Editor::apply_explorer_action) — are **buffer-local
//! default keymaps** installed by the `FileType nxdir` autocmd, not a special
//! `input()` branch. See docs/plans/2026-06-16-unify-special-buffer-kinds.md.

use super::*;
use crate::buffer::Buffer;
use std::path::Path;

impl Editor {
    /// Whether the current buffer is a directory listing (the file explorer).
    /// When true, [`Editor::key_context`] reports [`KeyContext::Explorer`] so the
    /// matcher routes normal-mode keys through the `explorer` keymap bucket
    /// ([`Editor::apply_explorer_action`]) rather than the editing state machine.
    pub(crate) fn is_explorer_buffer(&self) -> bool {
        self.buffer().dir().is_some()
    }

    /// Open `path` (a directory) as the file explorer. Reuses the current window
    /// in place when it holds a throwaway scratch buffer or is already an explorer
    /// (so `:e dir` from `[No Name]` and descending into a sub-directory both reuse
    /// the window, like netrw); otherwise opens a fresh listing buffer and switches
    /// to it, keeping the current buffer in the list. An unreadable directory fails
    /// loud — its OS error is echoed rather than leaving a blank buffer behind.
    pub(crate) fn enter_dir(&mut self, path: &Path) {
        // Off-tick (daemon session): the directory lives on the remote, so its
        // listing can't be read through the synchronous `host_fs` without blocking
        // the editor thread on the network. Set up the destination buffer — reuse
        // the window in place when it's a throwaway/explorer (netrw-style descend),
        // else a fresh listing buffer keeping the current one open — and enqueue an
        // off-tick fetch. The server reads the entries over `HostFsAsync` and fills
        // the buffer via `load_dir_into`; the old listing shows until then.
        if self.host_fs_offtick {
            let buf = if self.current_is_throwaway() || self.is_explorer_buffer() {
                self.cur_buffer()
            } else {
                let id = self.add_buffer(Buffer::named(path.to_path_buf()));
                self.switch_buffer(id);
                id
            };
            // Mark it `nxdir` now (before the off-tick fill) so the `FileType nxdir`
            // autocmd fires while the buffer is current — `load_dir_into` re-sets it
            // when the entries land, but by then the buffer is already announced.
            self.set_filetype(buf, "nxdir");
            self.enqueue_open(buf, path.to_path_buf());
            return;
        }

        // Path-based dedup: an already-open listing of the same directory is
        // reused (and reset to the top), matching the file open-or-switch path.
        let fs = self.host_fs.clone();
        let canon = fs.canonicalize(path).unwrap_or_else(|_| path.to_path_buf());
        if self.buffer().dir() != Some(canon.as_path()) {
            if let Some(id) = self.find_buffer_by_path(&canon) {
                self.switch_buffer(id);
                self.explorer_goto(0);
                return;
            }
        }
        match Buffer::from_dir(&canon, &*fs) {
            Ok(buf) => {
                if self.current_is_throwaway() || self.is_explorer_buffer() {
                    // Replace in place, preserving the buffer id (netrw reuses the
                    // window as you descend). The listing is freshly read from
                    // disk, so start a clean undo history rooted at it.
                    self.cursor = Cursor::default();
                    self.top = 0;
                    self.leftcol = 0;
                    let ob = self.cur_mut();
                    ob.buffer = buf;
                    ob.undo = UndoTree::new(&ob.buffer);
                    ob.saved_seq = Some(ob.undo.cur_seq());
                    ob.buffer.mark_resync();
                } else {
                    let id = self.add_buffer(buf);
                    self.switch_buffer(id);
                }
                // `filetype=nxdir` so the `FileType nxdir` autocmd installs the
                // explorer's buffer-local `<CR>`/`-` maps (the unified model).
                self.set_filetype(self.cur_buffer(), "nxdir");
                self.explorer_goto(0);
            }
            Err(e) => self.echo(e.to_string()),
        }
    }

    /// Apply a named `explorer` action, dispatched by a `FileType nxdir` buffer-local
    /// keymap (the default `<CR>`/`-` maps in `prelude/keymap.lua`, or a user
    /// override) while a directory-listing buffer is focused. `open` opens the entry
    /// under the cursor — a file is edited, a sub-directory is listed in place — and
    /// `up` goes to the parent. An unknown name fails loud per the no-silent-stub
    /// rule. Navigation (`j`/`k`/`gg`/`G`/`<C-d>`…) is ordinary normal-mode motion on
    /// the `nomodifiable` listing now, so only the activation keys are actions here.
    pub fn apply_explorer_action(&mut self, action: &str) -> Result<(), String> {
        self.message.clear();
        match action {
            "open" => self.explorer_open_entry(),
            "up" => self.explorer_up(),
            other => return Err(format!("unknown explorer action {other:?}")),
        }
        Ok(())
    }

    /// Move the listing selection to `line` (clamped), resting at column 0 and
    /// scrolling it into view. Used when (re)opening a listing to land on the top.
    fn explorer_goto(&mut self, line: usize) {
        self.cursor.line = line.min(self.last_line());
        self.cursor.col = 0;
        self.desired_col = 0;
        self.desired_eol = false;
        self.ensure_visible();
    }

    /// `<CR>` — open the entry under the cursor: descend into a sub-directory
    /// (listed in place), open a file in a new buffer, or go up on `../`. A no-op
    /// on a blank/garbled line.
    fn explorer_open_entry(&mut self) {
        let Some(dir) = self.buffer().dir().map(|p| p.to_path_buf()) else {
            return;
        };
        let line = self.buffer().line(self.cursor.line);
        let entry = line.trim_end_matches('/');
        if entry == ".." {
            self.explorer_up();
            return;
        }
        if entry.is_empty() {
            return;
        }
        let target = dir.join(entry);
        // The listing already encodes whether each entry is a directory (the trailing
        // `/` the listing builder appends), so off-tick we route on that — no remote
        // stat round-trip. Locally we stat directly, as before (a fresh `is_dir`).
        let is_dir = if self.host_fs_offtick {
            self.buffer().line(self.cursor.line).ends_with('/')
        } else {
            target.is_dir()
        };
        if is_dir {
            self.enter_dir(&target);
        } else {
            self.explorer_open_file(&target);
        }
    }

    /// `-` (and `<CR>` on `../`) — list the parent directory. At the filesystem
    /// root there is no parent, so the explorer stays put (as netrw does).
    fn explorer_up(&mut self) {
        let Some(dir) = self.buffer().dir().map(|p| p.to_path_buf()) else {
            return;
        };
        if let Some(parent) = dir.parent() {
            let parent = parent.to_path_buf();
            self.enter_dir(&parent);
        }
    }

    /// Open a file picked from the listing: switch to it if already open, else read
    /// it into a fresh buffer and switch. The explorer is then **destroyed** — it
    /// was a transient picker, so it does not linger in the buffer list or as the
    /// alternate (vim's netrw `bufhidden=wipe`). A read error is echoed and the
    /// listing is kept, so a failed open leaves you back in the explorer.
    fn explorer_open_file(&mut self, path: &Path) {
        let explorer = self.cur_buffer();
        // Open the picked file through the shared kernel — switch to it if already open,
        // else load a fresh buffer (off-tick over the wire in a daemon session, else from
        // local disk). A failed synchronous load is echoed; keep the listing so you land
        // back in the explorer rather than on a blank buffer.
        match self.open_buffer(path) {
            Some(id) => self.switch_buffer(id),
            None => return,
        }
        // Wipe the listing buffer now that we've left it. Forced because it is a
        // read-only listing that is never "modified", so the `E89` guard is moot.
        self.delete_buffer(explorer, true);
    }
}

File: crates/nxvim-lua/src/prelude/explorer.lua
-- nx.explorer visuals: highlighting for the in-window directory listing (vim's
-- netrw, `filetype=nxdir`). The listing itself — reading entries, `<CR>`/`-`
-- navigation — lives in the core (`editor/explorer.rs`), and its activation keys
-- are wired as buffer-local maps by the `FileType nxdir` autocmd in
-- `prelude/keymap.lua`. This file adds the *decoration* layer, and it does so the
-- way a plugin would: a viewport-scoped `nx.decor` provider (the new native
-- decoration API), not a special case baked into the renderer.
--
-- Loads AFTER `prelude/decor.lua` (it calls `nx.decor.provider`) — see the prelude
-- order in `runtime.rs`.

nx.explorer = nx.explorer or {}

-- The highlight groups the listing paints with. Concrete catppuccin-mocha colours
-- (the same convention as the `:Plugins` UI in `prelude/plugins_ui.lua`) so the
-- listing reads well on the default dark background even before a colorscheme
-- loads; a colorscheme can redefine these groups to match its own palette.
local HL = {
  NxDirDirectory = { fg = "#89b4fa", bold = true }, -- a sub-directory row (suffixed `/`)
  NxDirParent = { fg = "#94e2d5" }, -- the leading `../` up-entry
}
for name, spec in pairs(HL) do
  nx.hl.define(0, name, spec)
end

-- The provider: wake once per visible-range change of an `nxdir` window and colour
-- each directory row across the visible slice — the `../` up-entry one way, every
-- sub-directory (a line ending in `/`) another. File rows publish no mark, so they
-- render in the normal text colour, exactly like netrw. Re-listing in place
-- (descend / `-`) clears the old marks (the wholesale-rope-replace drops the
-- namespace) and re-dispatches this provider, so the colours always track the
-- current directory.
nx.decor.provider({
  name = "explorer",
  bufs = { filetype = { "nxdir" } },
  on_range = function(ctx, publish)
    local marks = {}
    for i, line in ipairs(ctx.lines) do
      local row = ctx.top + i - 1
      if line == "../" then
        marks[#marks + 1] = { row = row, col = 0, end_col = #line, hl = "NxDirParent" }
      elseif line:sub(-1) == "/" then
        marks[#marks + 1] = { row = row, col = 0, end_col = #line, hl = "NxDirDirectory" }
      end
    end
    publish(marks)
  end,
})

-- ============================================================================
-- The explorer itself — the in-window directory listing as a pure-Lua plugin.
--
-- This is the netrw mechanism, built on `nx.*`. A local `:e dir` is claimed by a
-- `BufReadCmd` handler (Primitive B of docs/plans/2026-06-25-explorer-lua-port.md),
-- which reads the entries with `nx.fs` and renders them into the opened buffer. A
-- *remote* directory (a daemon / web session) is filled by the server from the entries
-- its off-tick fetch already read (the `nx.fs` op and the file-open fetch are separate
-- legs, and a daemon may wire only the latter) — but in BOTH cases the result is the
-- same: a `nomodifiable`, `filetype=nxdir` buffer named for the directory.
--
-- Navigation is therefore **stateless**: it derives the directory from the buffer's
-- name and the entry from the row text (a trailing `/` marks a sub-directory), so it
-- works whoever filled the listing. `<CR>` / `<2-LeftMouse>` open the entry under the
-- cursor (descend in place, or edit a file), `-` goes to the parent; everything else is
-- plain normal-mode motion on the read-only buffer, and the decor provider above
-- colours the rows. The activation maps are installed by the `FileType nxdir` autocmd
-- (prelude/keymap.lua), which fires the moment either fill path sets the filetype.
-- ============================================================================

-- Escaping a path for use as a bare `:edit` argument is `nx.fname.escape`
-- (vimfn.lua — loaded after this module, referenced only at open time).
local function edit_escape(path)
  return nx.fname.escape(path)
end

-- Render a `nx.fs.readdir` result into listing lines — the Lua twin of the server's
-- `nxvim_core::dir_listing` (off-tick): `../` first, then sub-directories (suffixed
-- `/`), then files, each group sorted case-insensitively (netrw's default).
local function render(entries)
  local dirs, files = {}, {}
  for _, e in ipairs(entries) do
    if e.type == "directory" then
      dirs[#dirs + 1] = e.name
    else
      files[#files + 1] = e.name
    end
  end
  local byname = function(a, b)
    return a:lower() < b:lower()
  end
  table.sort(dirs, byname)
  table.sort(files, byname)
  local lines = { "../" }
  for _, name in ipairs(dirs) do
    lines[#lines + 1] = name .. "/"
  end
  for _, name in ipairs(files) do
    lines[#lines + 1] = name
  end
  return lines
end

-- Read `dir`'s entries (async) and render them into listing buffer `buf`: fill the
-- lines, lock the buffer (`nomodifiable`, and `modified = false` because the fill is a
-- read, not an edit — no `[+]`), and mark it `nxdir` (drives the decor provider + the
-- `FileType nxdir` activation maps). A failed read is surfaced loud and leaves the
-- buffer empty rather than a half listing. The LOCAL fill path (the `BufReadCmd`
-- handler); a remote directory is filled server-side (`nxvim_core::dir_listing`).
local function list_into(buf, dir)
  nx.async(function()
    local ok, entries = pcall(nx.await, nx.fs.readdir(dir))
    if not ok then
      local msg = type(entries) == "table" and entries.message or tostring(entries)
      return nx.notify("explorer: can't read " .. dir .. ": " .. msg, 4)
    end
    -- Fill, then lock. `nvim_buf_set_lines` refuses a `nomodifiable` buffer, so make it
    -- writable for the fill and lock it right after; clear `modified` since this fill is
    -- the file *read*, not an unsaved change.
    nx.bo[buf].modifiable = true
    vim.api.nvim_buf_set_lines(buf, 0, -1, false, render(entries))
    nx.bo[buf].modifiable = false
    nx.bo[buf].modified = false
    nx.bo[buf].filetype = "nxdir"
  end)()
end

-- The absolute directory a listing buffer represents — its name, canonicalised
-- (trailing slash trimmed except the root) so `../`/descent math is unambiguous however
-- it was spelled (`.`, a relative dir).
local function listing_dir(buf)
  local dir = vim.fn.fnamemodify(vim.api.nvim_buf_get_name(buf), ":p")
  if #dir > 1 then
    dir = dir:gsub("/$", "")
  end
  return dir
end

-- The row text under the cursor in listing buffer `buf` (the entry name, with a `/`
-- suffix for a sub-directory), or nil on a blank/garbled row.
local function entry_under_cursor(buf)
  local row = vim.api.nvim_win_get_cursor(0)[1] -- 1-based
  local line = vim.api.nvim_buf_get_lines(buf, row - 1, row, false)[1]
  if not line or line == "" then
    return nil
  end
  return line
end

-- Replace the listing in window-current buffer `cur` with a fresh `:edit` of `path`,
-- then wipe `cur`. `:edit` switches the window synchronously (the listing/file fill is
-- deferred — a directory is re-claimed by the `BufReadCmd` handler / refilled by the
-- server, a file read normally), so by the time the wipe runs `cur` is no longer
-- current. Net buffer count is unchanged — the netrw "reuse the window, don't strand
-- the old listing" behaviour (descend in place; opening a file destroys the listing).
local function replace_with(cur, path)
  -- The explorer's navigation math is absolute (an unambiguous `dir .. "/" .. name`), but
  -- the buffer the user lands on reads better named relative to cwd — netrw's model, and
  -- what `:edit relative/path` would have produced by hand. `:.` returns the path relative
  -- to cwd when it lives under it, and leaves an outside path (a parent above cwd, a remote
  -- `/virtual/...`) absolute. Relativise before escaping so the `:edit` arg is the final name.
  path = vim.fn.fnamemodify(path, ":.")
  vim.cmd("edit " .. edit_escape(path))
  vim.cmd("bwipeout! " .. cur)
end

-- `<CR>` / double-click: open the entry under the cursor. `../` (or the cursor on the
-- top row) goes up; any other entry is edited — a sub-directory re-lists, a file opens
-- (destroying the listing). A no-op on a stale/blank buffer.
function nx.explorer._open(buf)
  local line = entry_under_cursor(buf)
  if not line then
    return
  end
  if line == "../" then
    return nx.explorer._up(buf)
  end
  local dir = listing_dir(buf)
  local name = line:gsub("/$", "") -- a sub-directory row carries a trailing `/`
  local target = (dir == "/" and "/" or dir .. "/") .. name
  replace_with(buf, target)
end

-- `-` (and `<CR>` on `../`): list the parent directory. At the filesystem root the
-- parent is the root itself, so the explorer stays put (as netrw does).
function nx.explorer._up(buf)
  local dir = listing_dir(buf)
  local parent = vim.fn.fnamemodify(dir, ":h")
  if not parent or parent == "" or parent == dir then
    return
  end
  replace_with(buf, parent)
end

-- `nx.explorer.actions` — the public, rebindable activation actions (current-buffer
-- wrappers over the stateless `_open`/`_up`). A user/plugin extends or rebinds the
-- explorer with an ordinary buffer-local map in a `FileType nxdir` autocmd, e.g.
--   `nx.autocmd.create("FileType", { pattern = "nxdir", callback = function(a)`
--     `nx.keymap.set("n", "<C-j>", nx.explorer.actions.open, { buffer = a.buf })`
--   `end })`
-- The default `<CR>`/`-`/`<2-LeftMouse>` maps (prelude/keymap.lua) are these too.
nx.explorer.actions = nx.explorer.actions or {}
nx.explorer.actions.open = function()
  nx.explorer._open(vim.api.nvim_get_current_buf())
end
nx.explorer.actions.up = function()
  nx.explorer._up(vim.api.nvim_get_current_buf())
end

-- `nx.explorer.enable()`: turn the explorer on. Registers the `BufReadCmd` handler that
-- claims directory opens — `pattern = "*"` deciding per path via `args.isdir`, so it
-- claims directories and lets every file read fall through to the editor's default load
-- (netrw's exact model). Idempotent. Calling it makes a `BufReadCmd` handler exist,
-- which is what tells the core to defer a `:e dir` so this handler gets the chance to
-- claim it (see `Editor::should_defer_open`). Called at the bottom of this file, so the
-- explorer is on by default; a config could turn it off by clearing the autocmd.
function nx.explorer.enable()
  if nx.explorer._enabled then
    return
  end
  nx.explorer._enabled = true
  nx.autocmd.create("BufReadCmd", {
    pattern = "*",
    callback = function(args)
      if not args.isdir then
        return -- a file (or non-directory): decline, the default read fills it
      end
      -- Claim the read and fill asynchronously. The directory path is canonicalised to
      -- an absolute path (trailing slash trimmed, except the root) so `../` / descent
      -- math is unambiguous however it was spelled (`.`, a relative dir).
      local dir = vim.fn.fnamemodify(args.file, ":p")
      if #dir > 1 then
        dir = dir:gsub("/$", "")
      end
      list_into(args.buf, dir)
      return true
    end,
  })
end

-- The explorer is a built-in: on by default, the netrw replacement.
nx.explorer.enable()
